File: app/src/main/java/it/unical/mat/embasp/specializations/dlv2/DLV2AnswerSets.java
package it.unical.mat.embasp.specializations.dlv2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import it.unical.mat.embasp.languages.asp.AnswerSet;
import it.unical.mat.embasp.languages.asp.AnswerSets;

/** represents Dlv2's answersets */

public class DLV2AnswerSets extends AnswerSets {

	public DLV2AnswerSets(final String answersets) {
		super(answersets);
	}

	public DLV2AnswerSets(final String out, final String err) {
		super(out, err);
	}

	@Override
	protected void parse() {
		final boolean optimum = output.contains("OPTIMUM");
		final Pattern pattern = !optimum ? Pattern.compile("ANSWER\\r?\\n(.*)") : Pattern.compile("ANSWER\\r?\\n(.*)(\\r?\\nCOST (.+)\\r?\\nOPTIMUM)");

		final Matcher matcher = pattern.matcher(output);
		while (matcher.find()) {

			final Pattern patternAnswerSet = Pattern.compile("-?[a-z][A-Za-z0-9_]*(\\(.*?\\))?");
			final Matcher matcherAnswerSet = patternAnswerSet.matcher(matcher.group(1));
			final List<String> answerSetList = new ArrayList<>();
			while (matcherAnswerSet.find())
				answerSetList.add(matcherAnswerSet.group());

			if (optimum) {

				final Map<Integer, Integer> weightMap = new HashMap<>();
				try {
					final String[] split = matcher.group(3).split(" ");
					for (final String weightLevel : split) {
						final String[] weightLevelArray = weightLevel.split("@");
						weightMap.put(Integer.parseInt(weightLevelArray[1]), Integer.parseInt(weightLevelArray[0]));
					}
				} catch (final NumberFormatException e) {
					e.printStackTrace();
				}

				answersets.add(new AnswerSet(answerSetList, weightMap));

			} else
				answersets.add(new AnswerSet(answerSetList));
		}

	}

}

File: CSharp version/EmbASP/it/unical/mat/embasp/specializations/dlv2/DLV2AnswerSets.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace it.unical.mat.embasp.specializations.dlv2
{

	using AnswerSet = it.unical.mat.embasp.languages.asp.AnswerSet;
	using AnswerSets = it.unical.mat.embasp.languages.asp.AnswerSets;

	public class DLV2AnswerSets : AnswerSets
	{

		public DLV2AnswerSets(string answersets) : base(answersets)
		{
		}

		public DLV2AnswerSets(string @out, string err) : base(@out, err)
		{
		}

		protected internal override void parse()
		{
			bool optimum = output.Contains("OPTIMUM");
			string pattern = !optimum ? ("ANSWER\\r?\\n(.*)") : ("ANSWER\\r?\\n(.*)(\\r?\\nCOST (.+)\\r?\\nOPTIMUM)");

      Regex regex = new Regex(@pattern);
      Match matcher = regex.Match(output);

      while (matcher.Success)
			{

				string patternAnswerSet = ("-?[a-z][A-Za-z0-9_]*(\\(.*?\\))?");
        Regex regexAnswerSet = new Regex(patternAnswerSet);
				Match matcherAnswerSet = regexAnswerSet.Match(matcher.Groups[1].Value);
				IList<string> answerSetList = new List<string>();
				while (matcherAnswerSet.Success)
				{
					answerSetList.Add(matcherAnswerSet.Value);
          matcherAnswerSet = matcherAnswerSet.NextMatch();

        }

				if (optimum)
				{

					IDictionary<int, int> weightMap = new Dictionary<int, int>();
					try
					{
            string[] split = matcher.Groups[3].Value.Split(' '); //GroupNameFromNumber(3).Split(' ');
						foreach (String weightLevel in split)
						{
							string[] weightLevelArray = weightLevel.Split('@');
							weightMap[int.Parse(weightLevelArray[1])] = int.Parse(weightLevelArray[0]);
						}
					}
					catch (System.FormatException e)
					{
						Console.WriteLine(e.ToString());
						Console.Write(e.StackTrace);
					}

					answersets.Add(new AnswerSet(answerSetList, weightMap));

				}
				else
				{
					answersets.Add(new AnswerSet(answerSetList));
				}

        matcher = matcher.NextMatch();
      }

		}

	}

}